LearningGroup.Members is now an array of LearningGroupMember objects

brecriffs opened the pull request:
#### Summary
It will be useful to have an array of learning group member objects for each learning group (LearningGroup.Members), so as to have access to both the username and id of each member. This is in part inspired by requirements of the recommendations feature, but it also a better format to return the data in overall.

jaedoucette commented on app/user.go:
```diff
@@ -1691,9 +1691,28 @@ func (a *App) GetUserAnalytics(userId string, teamId string, name string) (*mode
+
+			for j, user := range usersInLearningGroup {
+				userData := strings.Split(user, ",")
+				membersArray[j] = &model.LearningGroupMember{Id: userData[0], Username: userData[1]}
```
Here, I don't see a great reason to use an array of pointers to LearningGroupMember structs rather than just using an array of LearningGroupMember structs? Maybe this makes JSON serialization easier later on? Otherwise, seems like the pointers aren't getting us anything except fragmentation and extra dereferencing?

brecriffs replied:
I guess I just opted for pointers whenever I was passing data around. My limited understanding of them, is that is can be better to pass a reference than the value.

jaedoucette replied:
That's not an unreasonable default. There's a tradeoff though. Two basic issues:

-> Using pointers means, all else being equal, two operations to get at data that would otherwise have cost one operation to access.
-> Allocating an array of pointers, and then making each point to an allocated object _can_ cause lots of small heap allocations. Heap allocation has a high overhead, so this will be much slower than the 1 big heap allocation caused to allocate an array of structs. 

In languages like C or C++ it's pretty easy to figure out when these tradeoffs are worth it. In Go, there's an extra complication, because the language specification makes no promises about when heap allocation will occur. Instead, it's going to perform "escape analysis", looking to see whether pointers to an allocated object "escape" the local context in which the object is created. In this case, they will (the array escapes, along with all the pointers in it), so I _think_ it'll do heap allocation of all the structs. My knowledge of this memory model is still far from perfect though...

Anyway, it's probably not a huge deal if it's a pain to change, but I _think_ this is going to be fairly expensive (like ~100x slower than it would otherwise be).

mlippert replied:
I also think this doesn't need to be an array of pointers.

One possible case where an array of pointers would be better than an array of the objects (structs) is when you were re-ordering or inserting/deleting often into the array. That doesn't happen with this array (I don't think).

brecriffs replied:
Oh, sorry, the main reason that I am in this habit is:
```
// An array of pointers to LearningGroupMember structs for every member of this learning group (channel)
	Members []*LearningGroupMember `json:"members"`
```
In the model, having the pointers allows it to be nil, which will be the case for a group having no members. I think the omit empty works for this as well. Maybe I will try that.

brecriffs replied:
So I think handling null values with pointers is only really necessary for sql results. I've changed this to not use pointers.

jaedoucette commented on model/user.go:
```diff
@@ -111,13 +111,25 @@ type LearningGroup struct {
+
 	// A comma delimited list of the usernames of the members of this learning group (channel)
-	Members *string `json:"members"`
+	MembersDelimitedString *string `json:"members_delimited_string"`
```
Do we still need the members delimited string if we have the learning group members?

brecriffs replied:
I was going to explain this, but I forgot! The LearningGroups struct represents a row for each result in the getUserLearningGroups query. I wasn't sure how to address that. Does it make more sense to have a struct for the query, and then another struct for the return data? I think having a useless property on the front-end is better than having two structs here.

jaedoucette replied:
Hmm. So the issue is that one part of the front end (dashboard) wants this delimited string, and another part (recs) wants the members structure? It's a bit ugly, but probably fine in the interests of time.

brecriffs replied:
No, they now both use the members structure.

mlippert replied:
Nope it absolutely makes sense to have a struct defined for the query and a different struct for what gets returned by the API. I just didn't want to make you change all that for the last PR.
Those really are 2 different things and they should be defined as such.

brecriffs replied:
Really? It will only have 1 different field, and the rest will be duplicates.

brecriffs replied:
Maybe it's better to use this technique. After we process MembersDelimitedString, set it to nil, and set the omit empty on it so it doesn't get passed to the front-end:
https://stackoverflow.com/questions/24216510/empty-or-not-required-struct-fields-in-golang

mlippert replied:
No, I still say that the struct returned by the API is a different entity from the struct returned internally by the query you wrote and they should be defined separately, and this is just the 1st crack in why they are different things, more will probably arise in the future. I was close to asking for the 2 struct definitions when this code was initially written but let it go since they will identical at that time.

The technique you're citing is not appropriate in my mind for this.

brecriffs replied:
Ok fair point. I think using pointers for null values in the struct is only required if the struct is a sql query result row...so it removes the need to use them in the return struct. I will make that adjustment.

jaedoucette commented on store/sqlstore/user_store.go:
```diff
@@ -1786,14 +1786,18 @@ func (us *SqlUserStore) GetUserLearningGroups(userId string, teamId string, lear
-				SELECT GROUP_CONCAT(Users.Username)
+				SELECT GROUP_CONCAT(
+									CONCAT(Users.id,',',Users.Username)
+									SEPARATOR ';'
```
This query continues to make me sad, but now is still not the time to fix it.

brecriffs replied:
Yeah..I agree.

jaedoucette reviewed (commented):
This looks basically good to me. There's one place where pointers are being used that doesn't really make sense to me. Otherwise, fine.

mlippert commented on app/user.go:
```diff
@@ -1691,9 +1692,44 @@ func (a *App) GetUserAnalytics(userId string, teamId string, name string) (*mode
 	} else {
+
+		userLearningGroupsQueryRows := userLearningGroups.Data.([]*model.LearningGroupQueryRow)
+		var userLearningGroups []*model.LearningGroup = make([]*model.LearningGroup, len(userLearningGroupsQueryRows))
```
Don't repurpose a variable. In this case I think the previous use of this variable `userLearningGroups` should be given a different name, perhaps `userLearningGroupsQueryResult`? Because it seems the right name for this definition.

mlippert commented on model/user.go:
```diff
@@ -115,13 +114,49 @@ type LearningGroup struct {
+	ChannelDisplayName string `json:"channel_display_name"`
+
 	// A comma delimited list of the usernames of the members of this learning group (channel)
 	Members *string `json:"members"`
```
I believe you've actually changed the value in Members (and commented on it elsewhere:
```
// Note: in a learning group query row, members is a double delimited string of users (; delimited) and
// each user's id and username (, delimited)
```

mlippert commented:
I pushed a commit addressing the minor code review comments I had.

brecriffs committed:
Sort learning group types by Prefix so they are returned in a consistent order

Since the learning group types are based on the names of the LMS LTI personal channels which
are in the config in a 'map' those names are in a random order. We want to make sure
that the learning group types are returned in a predictable order every time, so
we sort them by prefix.

brecriffs committed:
LearningGroup.Members is now an array of LearningGroupMember objects, with properties id and username

The query now returns Members semicolon(;) delimited and each of which contains an id and username
which are comma (,) delimited

mlippert commented:
squashed the commits for merging

mlippert commented:
**Release** [1.0.0-dev.11](https://github.com/rifflearning/mattermost-server/releases/tag/1.0.0-dev.11)
